/**
 * Copyright 2022 Huawei Technologies Co., Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
syntax = "proto3";
package mindspore.fl;

message CollectiveData {
  bytes data = 1;
}

message FLId {
  string fl_id = 1;
}

message DeviceMeta {
  string fl_name = 1;
  string fl_id = 2;
  uint64 data_size = 3;
  uint64 now_time = 4;
  uint64 eval_data_size = 5;
}

message FLIdToDeviceMeta {
  map<string, DeviceMeta> fl_id_to_meta = 1;
}

message UpdateModelThreshold {
  uint64 threshold = 1;
}

message ClientShares {
  map<string, SharesPb> client_secret_shares = 1;
}

message ClientKeys {
  map<string, KeysPb> client_keys = 1;
}

message ClientNoises {
  OneClientNoises one_client_noises = 1;
}

message OneClientNoises {
  repeated float noise = 1;
}

message ClientShareStr {
  string fl_id = 1;
  bytes share = 2;
  int32 index = 3;
}

message SharesPb {
  repeated ClientShareStr clientsharestrs = 1;
}

message KeysPb {
  repeated bytes key = 1;
  string timestamp = 2;
  int32 iter_num = 3;
  bytes ind_iv = 4;
  bytes pw_iv = 5;
  bytes pw_salt = 6;
  bytes signature = 7;
  repeated string certificate_chain = 8;
}

message Prime {
  bytes prime = 1;
}

message ClientListSign {
  map<string, bytes> client_list_sign = 1;
}

message KeyAttestation {
  map<string, string> key_attestations = 1;
}

message IterationSummaryMsg {
  string result = 1;
  uint64 startFLJob_total_client_num = 2;
  uint64 startFLJob_accept_client_num = 3;
  uint64 startFLJob_reject_client_num = 4;
  uint64 updateModel_total_client_num = 5;
  uint64 updateModel_accept_client_num = 6;
  uint64 updateModel_reject_client_num = 7;
  uint64 getModel_total_client_num = 8;
  uint64 getModel_accept_client_num = 9;
  uint64 getModel_reject_client_num = 10;
  float upload_loss = 11;
  uint64 participation_time_level1_num = 12;
  uint64 participation_time_level2_num = 13;
  uint64 participation_time_level3_num = 14;

  float metrics_loss = 15;
  float metrics_accuracy = 16;
  float upload_accuracy = 17;
  uint64 eval_data_size = 18;
  uint64 getResult_total_client_num = 19;
  uint64 getResult_accept_client_num = 20;
  uint64 getResult_reject_client_num = 21;
}

message UnsupervisedEvalItem {
  string fl_id = 1;
  repeated float eval_data = 2;
}